# Lab | Pandas

Objective: practice how to use the pandas library in Python for data analysis and manipulation.

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [2]:
# Your code here
import numpy as np
import pandas as pd 

In [3]:
file1 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')

In [5]:
display(file1.head())

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [28]:
print("Question : Identify the dimensions of the dataset by determining the number of rows and columns it contains.")
print("The number of rows and colums it contains is ", file1.shape)
print()
print("Therefore, the answer is 2 dimensions")

Question : Identify the dimensions of the dataset by determining the number of rows and columns it contains.
The number of rows and colums it contains is  (4008, 11)

Therefore, the answer is 2 dimensions


In [43]:
print("Question: Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable." , end="")
print(" You should also provide suggestions for fixing any incorrect data types.")
print()
print("The data types of each colums is as follows:")
print(file1.info())
print()
display(file1.head(3))
print("When you have look at the data above, the column of 'Customer Lifetime Value' currently has the type of 'object'.")
print("This can be 'float64' by removing '%'.")
print("In addition, 'Number of Open Complaints' also can be 'int64'.")

Question: Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.

The data types of each colums is as follows:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              10

Customer          ST GENDER Education Customer Lifetime Value   Income  \
0  RB50392  Washington    NaN    Master                     NaN      0.0   
1  QZ44356     Arizona      F  Bachelor              697953.59%      0.0   
2  AI49188      Nevada      F  Bachelor             1288743.17%  48767.0   

   Monthly Premium Auto Number of Open Complaints    Policy Type  \
0                1000.0                    1/0/00  Personal Auto   
1                  94.0                    1/0/00  Personal Auto   
2                 108.0                    1/0/00  Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247

When you have look at the data above, the column of 'Customer Lifetime Value' currently has the type of 'object'.
This can be 'float64' by removing '%'.
In addition, 'Number of Open Complaints' also can be 'int64'.


In [166]:
print("Question: Identify the number of unique values for each column and determine which columns", end="")
print("appear to be categorical.  You should also describe the unique values of each categorical column ", end="")
print("and the range of values for numerical columns, and give your insights.")
print()
print("The number of unique values for each column is as follows:")
print(file1.nunique())
print()
print("So, 'ST', 'GENDER','Education','Number of Open Complaints', 'Policy Type', and 'Vehicle Class' are categorical.")
print()
print("The unique values of each categorical column are as follows:")
print("ST ->",file1['ST'].unique())
print("GENDER ->",file1['GENDER'].unique())
print("Education ->",file1['Education'].unique())
print("Number of Open Complaints ->",file1['Number of Open Complaints'].unique())
print("Policy Type ->",file1['Policy Type'].unique())
print("Vehicle Class ->",file1['Vehicle Class'].unique())
print()
print("the range of values for numerical columns is as follows:")
print("Income ->" , file1['Income'].max()-file1['Income'].min())
print("Monthly Premium Auto ->" , file1['Monthly Premium Auto'].max()-file1['Monthly Premium Auto'].min())
print("Total Claim Amount ->" , file1['Total Claim Amount'].max()-file1['Total Claim Amount'].min())
print()
#My insights are below

Question: Identify the number of unique values for each column and determine which columnsappear to be categorical.  You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.

The number of unique values for each column is as follows:
Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

So, 'ST', 'GENDER','Education','Number of Open Complaints', 'Policy Type', and 'Vehicle Class' are categorical.

The unique values of each categorical column are as follows:
ST -> ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]
GENDER -> [nan 'F' 'M' 'Femal' 'Male

My insights:

1. There are more than two names for the same value in the columns of 'ST'and 'GENDER'. This has to be fixed.
2. Each number of 'Total Claim Amoun' could be rounded off to two decimal place.

In [113]:
print("Question: Compute summary statistics such as mean, median, mode, standard deviation,and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based  on these summary statistics.")
print()
display(file1.describe())
print()
#My conclusions are below this and written in 'Markdown' 


Question: Compute summary statistics such as mean, median, mode, standard deviation,and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based  on these summary statistics.



Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

Conclusion: 
1. Income: Looking at the standard deviation, its range is very wide.
2. Monthly Premium Auto: More than half of premium members pay less than 100.0.
3. Total Claim Amount: More than half of premium members pay more than 350.0.


In [111]:
print("Question: Compute summary statistics for categorical columns and providing your conclusions based on these statistics.")
display(file1['ST'].value_counts())
display(file1['GENDER'].value_counts())
display(file1['Education'].value_counts()) 
display(file1['Number of Open Complaints'].value_counts())
display(file1['Policy Type'].value_counts())
display(file1['Vehicle Class'].value_counts())
#My Conclusions are below this and written in 'Markdown'

Question: Compute summary statistics for categorical columns and providing your conclusions based on these statistics.


Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: ST, dtype: int64

F         457
M         413
Male       39
female     28
Femal      17
Name: GENDER, dtype: int64

Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: Education, dtype: int64

1/0/00    830
1/1/00    138
1/2/00     50
1/3/00     34
1/4/00     13
1/5/00      6
Name: Number of Open Complaints, dtype: int64

Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: Policy Type, dtype: int64

Four-Door Car    576
Two-Door Car     205
SUV              199
Sports Car        57
Luxury SUV        20
Luxury Car        14
Name: Vehicle Class, dtype: int64

Conslusions:
1. In the category of 'ST', 'Oregon' seems to have the majority, but, in fact, 'California' and 'Cali', both of which are identical, account for the great majority.
2. In the category of 'GENDER', at first, 'Femel' is mistyped. 'F','female'and 'femel' are all the same values, while 'M' and 'Male' are identical. These data can be labelled as M and F, or male and female. 
3. In the category of 'Education', 'Bachelor' and 'Bachelors' are the same, and they account for the greatest share of 'Education 
4. In the category of 'Number of Open Complaints', '1/0/00' accounts for the majority, but the data of this category has to changed into the data of cumerical columns.
5. In the category of 'Vehicle Class', its majority is 'Four-Door Car'(576), while the minority is 'Luxury car'(14). 

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [121]:
# Your code here
display(file1['ST'].value_counts().sort_values().head(5))

AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: ST, dtype: int64

### Exercise 2


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [122]:
# Your code here
#Review again the statistics for total claim amount to gain an understanding of the data.
file1['Total Claim Amount']

0          2.704934
1       1131.464935
2        566.472247
3        529.881344
4         17.269323
           ...     
4003            NaN
4004            NaN
4005            NaN
4006            NaN
4007            NaN
Name: Total Claim Amount, Length: 4008, dtype: float64

In [140]:
#To identify potential areas for improving customer retention and profitability, 
#we want to focus on customers with a high policy claim amount. 
#Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. 
#Create a pandas DataFrame object that contains information 
#about customers with a policy claim amount greater than the 75th percentile.

file1[ file1['Total Claim Amount'] >= 532.800000 ]


Total Claim Amount    532.8
Name: 0.75, dtype: float64


Customer          ST GENDER Education Customer Lifetime Value   Income  \
1     QZ44356     Arizona      F  Bachelor              697953.59%      0.0   
2     AI49188      Nevada      F  Bachelor             1288743.17%  48767.0   
17    OE15005        Cali    NaN   College              394524.16%  28855.0   
23    TZ98966      Nevada    NaN  Bachelor              245019.10%      0.0   
26    US89481  California    NaN  Bachelor              394637.21%      0.0   
...       ...         ...    ...       ...                     ...      ...   
1059  YG44474      Oregon      M   College             1401472.13%  54193.0   
1061  RY92647        Cali      F  Bachelor             1050677.17%      0.0   
1068  GS98873     Arizona      F  Bachelor              323912.47%  16061.0   
1069  CW49887  California      F    Master              462680.11%  79487.0   
1070  MY31220  California      F   College              899704.02%  54230.0   

      Monthly Premium Auto Number of Open Complaints     Policy Type  \
1                     94.0                    1/0/00   Personal Auto   
2                    108.0                    1/0/00   Personal Auto   
17                   101.0                    1/0/00   Personal Auto   
23                    73.0                    1/3/00  Corporate Auto   
26                   111.0                    1/0/00   Personal Auto   
...                    ...                       ...             ...   
1059                 117.0                    1/0/00  Corporate Auto   
1061                  92.0                    1/0/00   Personal Auto   
1068                  88.0                    1/0/00   Personal Auto   
1069                 114.0                    1/0/00    Special Auto   
1070                 112.0                    1/0/00   Personal Auto   

      Vehicle Class  Total Claim Amount  
1     Four-Door Car         1131.464935  
2      Two-Door Car          566.472247  
17              SUV          647.442031  
23    Four-Door Car          554.376763  
26    Four-Door Car          799.200000  
...             ...                 ...  
1059            SUV          720.752945  
1061  Four-Door Car          546.524896  
1068  Four-Door Car          633.600000  
1069            SUV          547.200000  
1070   Two-Door Car          537.600000  

[273 rows x 11 columns]

In [147]:
#Use DataFrame methods to calculate summary statistics about the high policy claim amount data
display(file1[['Total Claim Amount']].describe())


Total Claim Amount
count         1071.000000
mean           404.986909
std            293.027260
min              0.382107
25%            202.157702
50%            354.729129
75%            532.800000
max           2893.239678

### Exercise 3

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [149]:
# Your code here
file1['Policy Type'].value_counts().head(1)

Personal Auto    780
Name: Policy Type, dtype: int64

### Exercise 4

The sales team wants to know if customers with Personal Auto have a income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [164]:
# Your code here

df1 = file1[(file1["Policy Type"]) == "Personal Auto"]
df2 = file1[(file1["Policy Type"]) == "Corporate Auto"]

print("The inocome of customers with Personal Auto is ",df1['Income'].mean())
print("The inocome of customers with Corporate Auto is ",df2['Income'].mean())

The inocome of customers with PersonalAuto is  38180.69871794872
The inocome of customers with Corporate Auto is  41390.31196581197


Result:

Looking at the average, the customers with Coprorate Auto have more income than those with Personal Auto.